Tracking doc overhaul (#10471)
mlflow/mlflow: docs/source/tracking/artifacts-stores.rst

===============
Artifact Stores
===============

The artifact store is a core component in `MLflow Tracking <../index.html>`_ where MLflow stores (typicaly large) arifacts
for each run such as model weights (e.g. a pickled scikit-learn model), images (e.g. PNGs), model and data files (e.g. `Parquet <https://parquet.apache.org/>`_ file). 

> Note that metadata like parameters, metrics, and tags are stored in `backend store <backend-stores.html>`_, the other component of the MLflow Tracking.

Reviewer: ```suggestion
Note that metadata like parameters, metrics, and tags are stored in a `backend store <backend-stores.html>`_ (e.g., PostGres, MySQL, or MSSQL Database), the other component of the MLflow Tracking.
```
I think listing out examples of the backend stores might be helpful, just like you did with the artifact store examples :)

> Configure Artifact Store

Reviewer: ```suggestion
Configuring an Artifact Store
```
I think going with the gerund form fits in nicely with your other h2 headings below, as well as the instructional style of content

========================
MLflow by default stores artifacts in local ``./mlruns`` directory, but also supports various locations suitable for large data:

> AAmazon S3, Azure Blob Storage, Google Cloud Storage, SFTP server, and NFS. You can connect those remote storages via MLflow Tracking server.

Reviewer: ```suggestion
Amazon S3, Azure Blob Storage, Google Cloud Storage, SFTP server, and NFS. You can connect those remote storages via MLflow Tracking server.
```

> See :ref:`tracking server setup <tracking-server-artifact-store>` and the specific section for your storage in :ref:`supported storages <artifacts-store-supported-storages>` for 
> how to connect to the remote storage.

Reviewer: ```suggestion
See :ref:`tracking server setup <tracking-server-artifact-store>` and the specific section for your storage in :ref:`supported storages <artifacts-store-supported-storages>` for guidance on 
how to connect to your remote storage of choice.
```

.. _artifacts-stores-manage-access:

> Manage Access

Reviewer: ```suggestion
Managing Artifact Store Access
```

Gerund form reads more like the h2 heading on line 35

-------------
To allow the server and clients to access the artifact location, you should configure your cloud

> provider credentials as normal. For example, for S3, you can set the ``AWS_ACCESS_KEY_ID``

Reviewer: ```suggestion
provider credentials as you would for accessing them in any other capacity. For example, for S3, you can set the ``AWS_ACCESS_KEY_ID``
```

and ``AWS_SECRET_ACCESS_KEY`` environment variables, use an IAM role, or configure a default
profile in ``~/.aws/credentials``.

.. important::

>     Access credentials and configuration for the artifact storage location are configured *once during server initialization* in the place
>     of having users handle access credentials for artifact-based operations. Note that *all users who have access to the
>     Tracking Server in this mode will have access to artifacts served through this assumed role*.

Reviewer: ```suggestion
    Access credentials and configuration for the artifact storage location are configured **once during server initialization** in the place
    of having users handle access credentials for artifact-based operations. Note that **all users who have access to the
    Tracking Server in this mode will have access to artifacts served through this assumed role**.
```
This is pretty important. Let's go with `<strong>` vs `<i>` here

> Timeout

Reviewer: ```suggestion
Setting an access Timeout
```

-------
You can set an environment variable ``MLFLOW_ARTIFACT_UPLOAD_DOWNLOAD_TIMEOUT`` (in seconds) to configure the timeout for artifact uploads and downloads.

> Note that this is experimental feature, may be changed or removed.

Reviewer: ```suggestion
Note that this is an experimental feature which may be changed or removed.
```
Can we indicate what the default value is for this environment variable so that users don't have to look it up in source code?

Author: Makes sense! fyi apparently this one follows default timeout for each storage provider.

> Logging Artifacts Location

Reviewer: ```suggestion
Setting a Default Artifact Location for Logging
```

--------------------------

> MLflow automatically records ``artifact_uri`` property as a part of :py:class:`mlflow.entities.RunInfo`, so you can

Reviewer: ```suggestion
MLflow automatically records the ``artifact_uri`` property as a part of :py:class:`mlflow.entities.RunInfo`, so you can
```

retrieve the location of the artifacts for historical runs using the :py:func:`mlflow.get_artifact_uri` API. 

> Also, ``artifact_location`` is a property recorded on :py:class:`mlflow.entities.Experiment` for
> default location to store artifacts for all runs in this experiment.

Reviewer: ```suggestion
Also, ``artifact_location`` is a property recorded on :py:class:`mlflow.entities.Experiment` for setting the 
default location to store artifacts for all runs in a given experiment.
```

.. important::

  If you do not specify a ``--default-artifact-root`` or an artifact URI when creating the experiment
  (for example, ``mlflow experiments create --artifact-location s3://<my-bucket>``), the artifact root

>   is a path inside the file store. Typically this is not an appropriate location, as the client and

Reviewer: ```suggestion
  will be set as a path inside the local file store (the hard drive of the computer executing your run). Typically this is not an appropriate location, as the client and
```

  server probably refer to different physical locations (that is, the same path on different disks).

.. _artifacts-store-supported-storages:

> Supported storages

Reviewer: ```suggestion
Supported storage types for the Artifact Store
```

==================

Amazon S3 and S3-compatible storage
-----------------------------------

To store artifacts in S3 (whether on Amazon S3 or on an S3-compatible alternative, such as
`MinIO <https://min.io/>`_ or `Digital Ocean Spaces <https://www.digitalocean.com/products/spaces>`_), specify a URI of the form ``s3://<bucket>/<path>``. MLflow obtains
credentials to access S3 from your machine's IAM role, a profile in ``~/.aws/credentials``, or
the environment variables ``AWS_ACCESS_KEY_ID`` and ``AWS_SECRET_ACCESS_KEY`` depending on which of
these are available. For more information on how to set credentials, see
`Set up AWS Credentials and Region for Development <https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/setup-credentials.html>`_.

> To add S3 file upload extra arguments, set ``MLFLOW_S3_UPLOAD_EXTRA_ARGS`` to a JSON object of key/value pairs.

Reviewer: For searchability, we might want to wrap each of these examples and their explanation introductions in an h3 header ``^^^^``. These examples are pretty important.

Author: True, will make them distinct subsections.

.. code-block:: bash

  export AWS_DEFAULT_REGION=my_region

.. warning::

        The MLflow tracking server utilizes specific reserved keywords to generate a qualified path. These environment configurations, if present in the client environment, can create path resolution issues.
        For example, providing ``--default-artifact-root $MLFLOW_S3_ENDPOINT_URL`` on the server side **and** ``MLFLOW_S3_ENDPOINT_URL`` on the client side will create a client path resolution issue for the artifact storage location.
        Upon resolving the artifact storage location, the MLflow client will use the value provided by ``--default-artifact-root`` and suffixes the location with the values provided in the environment variable  ``MLFLOW_S3_ENDPOINT_URL``.
        Depending on the value set for the environment variable ``MLFLOW_S3_ENDPOINT_URL``, the resulting artifact storage path for this scenario would be one of the following invalid object store paths:  ``https://<bucketname>.s3.<region>.amazonaws.com/<key>/<bucketname>/<key>`` or  ``s3://<bucketname>/<key>/<bucketname>/<key>``.

>         To prevent path parsing issues, ensure that reserved environment variables are removed (``unset``) from client environments.

Reviewer: ```suggestion
        To prevent path parsing issues, **ensure that reserved environment variables are removed (``unset``) from client environments**.
```
emphasis

> Complete list of configurable values for an S3 client is available in `boto3 documentation <https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#configuration>`_.

Reviewer: ```suggestion
The complete list of configurable parameters for an S3 client is available in the `boto3 documentation <https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#configuration>`_.
```

Azure Blob Storage
------------------

To store artifacts in Azure Blob Storage, specify a URI of the form
``wasbs://<container>@<storage-account>.blob.core.windows.net/<path>``.

> MLflow expects Azure Storage access credentials in the

Reviewer: ```suggestion
MLflow expects that your Azure Storage access credentials are located in the
```

> ``AZURE_STORAGE_CONNECTION_STRING``, ``AZURE_STORAGE_ACCESS_KEY`` environment variables

Reviewer: ```suggestion
``AZURE_STORAGE_CONNECTION_STRING`` and ``AZURE_STORAGE_ACCESS_KEY`` environment variables
```

or having your credentials configured such that the `DefaultAzureCredential()
<https://docs.microsoft.com/en-us/python/api/overview/azure/identity-readme?view=azure-python>`_. class can pick them up.
The order of precedence is:

#. ``AZURE_STORAGE_CONNECTION_STRING``
#. ``AZURE_STORAGE_ACCESS_KEY``
#. ``DefaultAzureCredential()``

> You must set one of these options on both your client application and your MLflow tracking server.

Reviewer: ```suggestion
You must set one of these options on **both your client application and your MLflow tracking server**.
```

The URI may optionally include a password for logging into the server, e.g. ``ftp://user:pass@host/path/to/directory``

SFTP Server
-----------

To store artifacts in an SFTP server, specify a URI of the form ``sftp://user@host/path/to/directory``.
You should configure the client to be able to log in to the SFTP server without a password over SSH (e.g. public key, identity file in ssh_config, etc.).

The format ``sftp://user:pass@host/`` is supported for logging in. However, for safety reasons this is not recommended.

When using this store, ``pysftp`` must be installed on both the server and the client. Run ``pip install pysftp`` to install the required package.

NFS
---

To store artifacts in an NFS mount, specify a URI as a normal file system path, e.g., ``/mnt/nfs``.
This path must be the same on both the server and the client -- you may need to use symlinks or remount
the client in order to enforce this property.

HDFS
----

To store artifacts in HDFS, specify a ``hdfs:`` URI. It can contain host and port: ``hdfs://<host>:<port>/<path>`` or just the path: ``hdfs://<path>``.

There are also two ways to authenticate to HDFS:

- Use current UNIX account authorization

> - Kerberos credentials using following environment variables:

Reviewer: ```suggestion
- Kerberos credentials using the following environment variables:
```

.. code-block:: bash

  export MLFLOW_KERBEROS_TICKET_CACHE=/tmp/krb5cc_22222222
  export MLFLOW_KERBEROS_USER=user_name_to_use

Most of the cluster contest settings are read from ``hdfs-site.xml`` accessed by the HDFS native
driver using the ``CLASSPATH`` environment variable.

> The used HDFS driver is ``libhdfs``.

Reviewer: ```suggestion
The HDFS driver that is used is ``libhdfs``.
```
